<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Quantitative DNA sequencing for chromosomal aberrations</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Quantitative DNA sequencing for chromosomal aberrations
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;QDNAseqmod&rsquo; version 1.21.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="QDNAseq-package.html">QDNAseq-package</a></td>
<td>Package QDNAseq</td></tr>
<tr><td style="width: 25%;"><a href="addPhenodata.html">addPhenodata</a></td>
<td>Adds phenotype data from a file to a QDNAseqReadCounts or a QDNAseqCopyNumbers object</td></tr>
<tr><td style="width: 25%;"><a href="applyFilters.html">applyFilters</a></td>
<td>Adjusts the filtering on which bins are used</td></tr>
<tr><td style="width: 25%;"><a href="applyFilters.html">applyFilters-method</a></td>
<td>Adjusts the filtering on which bins are used</td></tr>
<tr><td style="width: 25%;"><a href="binReadCounts.html">binReadCounts</a></td>
<td>Calculate binned read counts from a set of BAM files</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqSignals.html">bpend-method</a></td>
<td>A parent class for containers of QDNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqSignals.html">bpstart-method</a></td>
<td>A parent class for containers of QDNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="createBins.html">calculateBlacklist</a></td>
<td>Builds bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="createBins.html">calculateBlacklistByRegions</a></td>
<td>Builds bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="createBins.html">calculateMappability</a></td>
<td>Builds bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="callBins.html">callBins</a></td>
<td>Call aberrations from segmented copy number data</td></tr>
<tr><td style="width: 25%;"><a href="callBins.html">callBins-method</a></td>
<td>Call aberrations from segmented copy number data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqSignals.html">chromosomes-method</a></td>
<td>A parent class for containers of QDNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="compareToReference.html">compareToReference</a></td>
<td>Divide binned read counts with those of reference samples</td></tr>
<tr><td style="width: 25%;"><a href="compareToReference.html">compareToReference-method</a></td>
<td>Divide binned read counts with those of reference samples</td></tr>
<tr><td style="width: 25%;"><a href="correctBins.html">correctBins</a></td>
<td>Correct binned read counts for GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="correctBins.html">correctBins-method</a></td>
<td>Correct binned read counts for GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="createBins.html">createBins</a></td>
<td>Builds bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseq-defunct.html">downloadBinAnnotations</a></td>
<td>Defunct functions in package 'QDNAseq'</td></tr>
<tr><td style="width: 25%;"><a href="estimateCorrection.html">estimateCorrection</a></td>
<td>Estimate correction to read counts for GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="estimateCorrection.html">estimateCorrection-method</a></td>
<td>Estimate correction to read counts for GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="exportBins.html">exportBins</a></td>
<td>Exports to a file</td></tr>
<tr><td style="width: 25%;"><a href="exportBins.html">exportBins-method</a></td>
<td>Exports to a file</td></tr>
<tr><td style="width: 25%;"><a href="frequencyPlot.html">frequencyPlot</a></td>
<td>Plot copy number aberration frequencies</td></tr>
<tr><td style="width: 25%;"><a href="frequencyPlot.html">frequencyPlot-method</a></td>
<td>Plot copy number aberration frequencies</td></tr>
<tr><td style="width: 25%;"><a href="getBinAnnotations.html">getBinAnnotations</a></td>
<td>Gets bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="highlightFilters.html">highlightFilters</a></td>
<td>Highlights data points in a plotted profile to evaluate filtering</td></tr>
<tr><td style="width: 25%;"><a href="highlightFilters.html">highlightFilters-method</a></td>
<td>Highlights data points in a plotted profile to evaluate filtering</td></tr>
<tr><td style="width: 25%;"><a href="isobarPlot.html">isobarPlot</a></td>
<td>Plot median read counts as a function of GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="isobarPlot.html">isobarPlot-method</a></td>
<td>Plot median read counts as a function of GC content and mappability</td></tr>
<tr><td style="width: 25%;"><a href="createBins.html">iterateResiduals</a></td>
<td>Builds bin annotation data for a particular bin size</td></tr>
<tr><td style="width: 25%;"><a href="LGG150.html">LGG150</a></td>
<td>LGG150 chromosomes 7-10</td></tr>
<tr><td style="width: 25%;"><a href="makeCgh.html">makeCgh</a></td>
<td>Constructs a 'cghRaw', 'cghSeg', or 'cghCall' object</td></tr>
<tr><td style="width: 25%;"><a href="makeCgh.html">makeCgh-method</a></td>
<td>Constructs a 'cghRaw', 'cghSeg', or 'cghCall' object</td></tr>
<tr><td style="width: 25%;"><a href="noisePlot.html">noisePlot</a></td>
<td>Plot noise as a function of sequence depth</td></tr>
<tr><td style="width: 25%;"><a href="noisePlot.html">noisePlot-method</a></td>
<td>Plot noise as a function of sequence depth</td></tr>
<tr><td style="width: 25%;"><a href="normalizeBins.html">normalizeBins</a></td>
<td>Normalizes binned read counts</td></tr>
<tr><td style="width: 25%;"><a href="normalizeBins.html">normalizeBins-method</a></td>
<td>Normalizes binned read counts</td></tr>
<tr><td style="width: 25%;"><a href="normalizeSegmentedBins.html">normalizeSegmentedBins</a></td>
<td>Normalize segmented bins</td></tr>
<tr><td style="width: 25%;"><a href="normalizeSegmentedBins.html">normalizeSegmentedBins-method</a></td>
<td>Normalize segmented bins</td></tr>
<tr><td style="width: 25%;"><a href="plot.html">plot</a></td>
<td>Plot copy number profile</td></tr>
<tr><td style="width: 25%;"><a href="plot.html">plot-method</a></td>
<td>Plot copy number profile</td></tr>
<tr><td style="width: 25%;"><a href="poolRuns.html">poolRuns</a></td>
<td>Pools binned read counts across samples</td></tr>
<tr><td style="width: 25%;"><a href="poolRuns.html">poolRuns-method</a></td>
<td>Pools binned read counts across samples</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseq-package.html">QDNAseq</a></td>
<td>Package QDNAseq</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqCopyNumbers.html">QDNAseqCopyNumbers</a></td>
<td>Container for QDNAseq read count data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqCopyNumbers.html">QDNAseqCopyNumbers-class</a></td>
<td>Container for QDNAseq read count data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqReadCounts.html">QDNAseqReadCounts</a></td>
<td>Container for QDNAseq read count data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqReadCounts.html">QDNAseqReadCounts-class</a></td>
<td>Container for QDNAseq read count data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqSignals.html">QDNAseqSignals</a></td>
<td>A parent class for containers of QDNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="QDNAseqSignals.html">QDNAseqSignals-class</a></td>
<td>A parent class for containers of QDNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="segmentBins.html">segmentBins</a></td>
<td>Segments normalized copy number data</td></tr>
<tr><td style="width: 25%;"><a href="segmentBins.html">segmentBins-method</a></td>
<td>Segments normalized copy number data</td></tr>
<tr><td style="width: 25%;"><a href="smoothOutlierBins.html">smoothOutlierBins</a></td>
<td>Smooth outlier bins after normalization</td></tr>
<tr><td style="width: 25%;"><a href="smoothOutlierBins.html">smoothOutlierBins-method</a></td>
<td>Smooth outlier bins after normalization</td></tr>
</table>
</body></html>
